flowmachine/flowmachine/core/model_result.py

```diff
@@ -150,25 +152,50 @@ def to_sql(
 
         def do_query() -> ModelResult:
             logger.debug("Getting storage lock.")
-            with rlock(self.redis, self.md5):
+            q_state_machine = QueryStateMachine(self.redis, self.md5)
+            current_state, this_thread_is_owner = q_state_machine.execute()
+            if this_thread_is_owner:
                 logger.debug("Obtained storage lock.")
                 con = self.connection.engine
-                if force:
-                    self.invalidate_db_cache(name, schema=schema)
                 try:
                     with con.begin():
                         logger.debug("Using pandas to store.")
                         self._df.to_sql(name, con, schema=schema, index=False)
+                        # Mark as finished before writing meta to avoid getting blocked when
+                        # meta calls _make_query
+                        q_state_machine.finish()
                         if schema == "cache":
                             self._db_store_cache_metadata(compute_time=self._runtime)
+
                 except AttributeError:
                     logger.debug(
                         "No dataframe to store, presumably because this"
                         " was retrieved from the db."
                     )
-            logger.debug("Released storage lock.")
+            elif q_state_machine.is_executing:
+                logger.debug(
+                    f"Model result '{self.md5}' being written elsewhere, waiting for it to finish."
+                )
+                while q_state_machine.is_executing:
+                    sleep(5)
```

Comment:
Could make this poll frequency an optional kwarg to `do_query()`? Maybe not important, but somehow it feels odd to have an arbitrary hard-coded number of seconds here that the call will be blocked for without the user being able to change it.